user - add password expiration parameters

pushkarkumar15 opened the pull request:
##### SUMMARY
<!--- Describe the change below, including rationale and design decisions -->
Add `password_expire_max` and `password_expire_min` to the `user` module.

Fixes #68775 

<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->

##### ISSUE TYPE
<!--- Pick one below and delete the rest -->
- Feature Pull Request


##### COMPONENT NAME
<!--- Write the short name of the module, plugin, task or feature below -->
`lib/ansible/modules/user.py`

##### ADDITIONAL INFORMATION
 Two parameters added in user module as shown in component section.
 It is used for setting a minimum and maximum password expire date.
<!--- Include additional information to help people understand the change here -->
<!--- A step-by-step reproduction of the problem is helpful if there is no related issue -->

<!--- Paste verbatim command output below, e.g. before and after your change -->
```paste below

```

Commit:
updated user.py file

jimi-c commented on lib/ansible/modules/user.py:
```diff
@@ -186,7 +186,7 @@
             - An expiry time for the user in epoch, it will be ignored on platforms that do not support this.
             - Currently supported on GNU/Linux, FreeBSD, and DragonFlyBSD.
-            - Since Ansible 2.6 you can remove the expiry time by specifying a negative value.
+            - Since Ansible 2.6 you can remove the expiry time specify a negative value.
```
The grammar was correct before, this line should be undone.

jimi-c replied:
Beyond this minor nit-pick, this PR looks ok.

samdoran commented on changelogs/fragments/69531_user_password_expire.yml:
```diff
@@ -0,0 +1,2 @@
+minor_changes:
+             - added new feature of users password expire max and min.
```
```suggestion
  - user - add new parameters ``password_expire_max `` and ``password_expire_min`` for controlling password expiration (https://github.com/ansible/ansible/issues/68775)
```

samdoran commented on lib/ansible/modules/user.py:
```diff
@@ -397,6 +416,16 @@
   type: int
   sample: 1044
+password_expire_max:
+  description: Maximum number of days between password change
```
```suggestion
  description: Maximum number of days during which a password is valid.
```

samdoran reviewed (changes requested):
Please add integration tests for these new parameters.

This only adds password max and min age for Linux. This should be added to other platforms as well where possible. If this will not work on all platforms, a note needs to be added to the parameter description stating on which platform these parameters work.

pushkarkumar15 committed:
Apply suggestions from code review

grammatical error resolved

Co-authored-by: Sam Doran <sdoran@redhat.com>

samdoran commented on lib/ansible/modules/user.py:
```diff
@@ -560,6 +591,26 @@ def backup_shadow(self):
+        cmd.append(self.password_expire_max)
+        cmd.append(self.name)
+
+        return self.execute_command(cmd)
```
Rather than running this command every time, we need to check the current value on the system and compare it to the value from the task.

The `user_password()` and `parse_shadow_file()` methods should be extended to return this information, then use that to determine whether or not to run these commands.

samdoran commented on lib/ansible/modules/user.py:
```diff
@@ -236,6 +236,17 @@
         version_added: "2.8"
+    password_expire_max:
+        description:
+            - Maximum number of days between password change.
```
It looks like this only works on Linux. We should make this work on platforms that support it, or put a note here that it only works on Linux.

pushkarkumar15 replied:
okay working on that .

samdoran reviewed (changes requested):
Please add integration tests.

samdoran commented on lib/ansible/modules/user.py:
```diff
@@ -236,6 +236,17 @@
+        description:
+            - Maximum number of days between password change.
+        type: int
+        version_added: "2.10"
```
```suggestion
        version_added: "2.11"
```

samdoran commented on lib/ansible/modules/user.py:
```diff
@@ -236,6 +236,17 @@
+        description:
+            - Minimum number of days between password change.
+        type: int
+        version_added: "2.10"
```
```suggestion
        version_added: "2.11"
```

samdoran commented:
Integration tests are for the `user` module are Ansible tasks located at `test/integration/targets/user/tasks/main.yml`. Add tasks that use this new parameter and verify they make the correct changes to the system.

pushkarkumar15 commented:
working on that

Commit:
integration test

pushkarkumar15 commented:
added integration test

samdoran commented on test/integration/targets/user/tasks/main.yml:
```diff
@@ -186,6 +186,47 @@
+        that:
+          - pass_min_2_0 is changed
+          - pass_min_2_1 is not changed
+  when: ansible_facts.system == 'Debian'
```
This is a great start. You can use the `getent` module to look at the shadow file and verify it got the correct values.
```yaml
- name: Get shadow data for ansibulluser
  getent:
    database: shadow
    key: ansibulluser
  
- name: Ensure password expiration was set properly
  assert:
    that:
       - getent_shadow[2] == '5'
       - getent_shadow[3] == '10'
```

pushkarkumar15 replied:
ok working on it.

samdoran commented on test/integration/targets/user/tasks/main.yml:
```diff
@@ -186,6 +186,63 @@
+      getent:
+        database: shadow
+        key: ansibulluser
+      register: getent_shadow[]
```
```suggestion
```

There's no need to register a variable. The `getent` module creates a `getent_[database name]` fact automatically.

samdoran commented on test/integration/targets/user/tasks/main.yml:
```diff
@@ -186,6 +186,63 @@
+        that:
+          - getent_shadow[2] == '5'
+          - getent_shadow[3] == '10'
+  when: ansible_facts.system == 'Debian'
```
We'll need to do this on all Linux distributions. Here's an example: https://github.com/ansible/ansible/blob/7525503512369b4fe0b028fb5884aedb07900764/test/integration/targets/user/tasks/main.yml#L124-L135

pushkarkumar15 commented:
Hi @samdoran  When I added the OS flavours for the testcases, your build is failing with "value_specified_in_no_log_parameter" for the updated values for minimum and maximum days. The actual values are not being updated, but inplace of the value you can see "value_specified_in_no_log_parameter". You tried by running the testcase as root but still it didn't work.

samdoran commented on test/integration/targets/user/tasks/main.yml:
```diff
@@ -186,6 +186,62 @@
+        that:
+          - getent_shadow[2] == '5'
+          - getent_shadow[3] == '10'
+  when: ansible_facts.system == '*'
```
This will never match anything so these tests are skipped.

samdoran replied:
It's also a good idea to put the conditional at the top of the block for easier reading.

samdoran commented:
The way `no_log` works is by replacing _all_ occurrences of the value with "value_specified_in_no_log_parameter". I understand why these values should be hidden, but very short and non-unique strings specified as `no_log` values will have unintended consequences and should be avoided.

Commit:
nolog false

Commit:
added os family

Commit:
Update tests

mkrizek commented:
!waffling needs_revision

mkrizek commented:
!waffling core_review

pushkarkumar15 commented:
@samdoran added test case kindly review .

pushkarkumar15 commented:
@samdoran kindly review the changes I have made.

pushkarkumar15 commented:
@samdoran kindly review the changes I have made.

pushkarkumar15 commented:
hi @samdoran kindly review my PR

samdoran commented:
/azp run

samdoran committed:
Fix bad merge

samdoran committed:
Fix tests to match organizational changes

samdoran committed:
Minor docs change

samdoran committed:
Remove space in changelog

samdoran committed:
Remove unnecessary methods

There is no benefit to having shim methods. If the behavior needs to change for a certain
platform, it can be overridden in the sub class.

samdoran committed:
Move methods to a place they make a bit more sense.

samdoran committed:
Fixup examples in docs

samdoran committed:
Rename test tasks file

It's testing min and max values not just max

pushkarkumar15 commented:
Thanks !! for merging the PR.
